server/lib/chat/chat-service.ts

```diff
@@ -490,43 +533,44 @@ export default class ChatService {
     })
   }
 
-  async getChannelInfo(channelId: SbChannelId, userId: SbUserId): Promise<ChannelInfo> {
-    const channelInfo = await getChannelInfo(channelId)
+  async getChannelInfo(channelId: SbChannelId, userId: SbUserId): Promise<GetChannelInfoResponse> {
+    const [channelInfo, userChannelEntry] = await Promise.all([
+      getChannelInfo(channelId),
+      getUserChannelEntryForUser(userId, channelId),
+    ])
+    const isUserInChannel = Boolean(userChannelEntry)
 
     if (!channelInfo) {
       throw new ChatServiceError(ChatServiceErrorCode.ChannelNotFound, 'Channel not found')
     }
 
-    let userCount
-    if (!channelInfo.private || this.state.users.get(userId)?.has(channelId)) {
-      userCount = channelInfo.userCount
-    }
-
+    const channelSummary = toChannelSummary(channelInfo)
     return {
-      id: channelInfo.id,
-      name: channelInfo.name,
-      private: channelInfo.private,
-      official: channelInfo.official,
-      userCount,
+      channelInfo: channelSummary.channelInfo,
+      detailedChannelInfo:
+        !channelInfo.private || isUserInChannel ? channelSummary.detailedChannelInfo : undefined,
     }
   }
 
-  async getChannelInfos(channelIds: SbChannelId[], userId: SbUserId): Promise<ChannelInfo[]> {
-    const channelInfos = await getChannelInfos(channelIds)
+  async getChannelInfos(
+    channelIds: SbChannelId[],
+    userId: SbUserId,
+  ): Promise<GetBatchedChannelInfosResponse> {
+    const [channelInfos, userChannelEntries] = await Promise.all([
+      getChannelInfos(channelIds),
+      getUserChannelEntriesForUser(userId, channelIds),
+    ])
 
-    return channelInfos.map(channel => {
-      let userCount
-      if (!channel.private || this.state.users.get(userId)?.has(channel.id)) {
-        userCount = channel.userCount
-      }
-      return {
-        id: channel.id,
-        name: channel.name,
-        private: channel.private,
-        official: channel.official,
-        userCount,
-      }
-    })
+    const userJoinedChannelsSet = new global.Set(userChannelEntries.map(e => e.channelId))
+
+    return {
+      channelInfos: channelInfos.map(channel => toChannelSummary(channel).channelInfo),
+      detailedChannelInfos: channelInfos
+        .filter(channel => {
+          return !channel.private || userJoinedChannelsSet.has(channel.id)
+        })
+        .map(channel => toChannelSummary(channel).detailedChannelInfo),
```

Comment:
Doing this conversion every time you need it feels incredibly wasteful. Just write the code to pull the fields you need? Write a for loop to build these arrays iteratively? I dunno, something other than this

Same thing with `joinedChannelInfo` above also applies